gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/iapi/db/ConsistencyChecker.java

```diff
@@ -504,170 +530,267 @@ public static boolean checkTable(String schemaName, String tableName)
-              td.getSchemaName(), 
-              td.getName(),
-              baseRL.toString(),
-              cd.getColumnName(),
-              RowUtil.toString(baseObjectArray),
-              baseRow.toString()
-              );
-        }
-      }
-      */
-    }
-    // now check the total number of keys in the index
-    int numIndexEntries = 0;
-    if (indexScan != null) {
-      indexScan.close();
-      indexScan = null;
-    }
-    indexScan = tc.openScan(indexCD.getConglomerateNumber(), false, 0,
-        TransactionController.MODE_TABLE,
-        TransactionController.ISOLATION_SERIALIZABLE,
-        (FormatableBitSet)null,
-        null, // null startKeyValue for a full scan
-        ScanController.GE, 
-        null, // qualifier
-        null, ScanController.GT, null);
-    DataValueDescriptor[] indexRowArray;
-    while (indexScan.next()) {
-      indexRowArray = indexRow.getRowArray();
-      indexScan.fetch(indexRow);
-      key = new CompositeRegionKey(getNewArray(indexRowArray, baseColumns,
-          false));
-      if (!tableIndexKeys.contains(key)) {
-        throw StandardException.newException(
-            SQLState.LANG_INCONSISTENT_ROW_LOCATION, td.getQualifiedName(),
-            indexCD.getConglomerateName(), key.toString(), indexRow.toString());
-      }
-      ++numIndexEntries;
-    }
-    if (numIndexEntries != tableIndexKeys.size()) {
-      throw StandardException.newException(
-          SQLState.LANG_INDEX_ROW_COUNT_MISMATCH,
-          indexCD.getConglomerateName(), td.getSchemaName(), td.getName(),
-          numIndexEntries, tableIndexKeys.size());
-    }
-    if (indexScan != null) {
-      indexScan.close();
-      indexScan = null;
-    }
-    scan.close();
-    scan = null;
-  }
+			if ((numRows % 1000) == 0) {
+				checkTableKeysInGlobalIndexRegion(td, indexCD, globalIndexRegion,
+						baseRow, searchCondition, tableIndexKeys);
+				tableIndexKeys.clear();
+			}
+		}
+
+		if (tableIndexKeys.size() > 0) {
+			checkTableKeysInGlobalIndexRegion(td, indexCD, globalIndexRegion,
+					baseRow, searchCondition, tableIndexKeys);
+			tableIndexKeys.clear();
+		}
+
+		// now check the total number of keys in the index
+		int numIndexEntries = 0;
+		if (indexScan != null) {
+			indexScan.close();
+			indexScan = null;
+		}
+		if (indexScan != null) {
+			indexScan.close();
+			indexScan = null;
+		}
+		scan.close();
+		scan = null;
+	}
+
+	private static void checkTableKeysInGlobalIndexRegion(TableDescriptor td,
+			ConglomerateDescriptor indexCD, LocalRegion globalIndexRegion,
+			ExecRow baseRow, DataValueDescriptor[] searchCondition,
+			HashSet<Object> tableIndexKeys) throws StandardException {
+		Map globalIndexEntriesMap = globalIndexRegion.getAll(tableIndexKeys);
+		for (Object o : tableIndexKeys) {
+			if (globalIndexEntriesMap.get(o) == null) {
+				DataValueDescriptor keyColumns[] = new DataValueDescriptor[searchCondition.length];
+				if (searchCondition.length == 1) {
```

Comment:
May be later we can add routing object correctness with the actual column's routing object using the partitoin resolver. We should also have a ticket to trac a lighter overall check for index consistency which will just check the sizes and come back.